Issue/2711 Add downloadable files to a product

hichamboushaba opened the pull request:
Fixes #2711 

**Description**
This PR add the implementation of adding downloadable files flow to products, the list of important changes:

- Adds an entry to the "Add more details" sheet to allow adding downloadable files if the product wasn't downloadable, or didn't have any downloadable files before.
- Created [MediaFilesRepository](https://github.com/woocommerce/woocommerce-android/blob/8ee07126e286d8d8d3c7498bbcd1347462e449f6/WooCommerce/src/main/kotlin/com/woocommerce/android/media/MediaFilesRepository.kt) to expose a coroutine for uploading files.
- Modified the viewstates to allow displaying a progress dialog when a file is being uploaded.
- Made some changes to the WPMediaGalleryView to allow single select mode.
- Made some changes to WPMediaPickerFragment to allow defining requestCode in navArgs, and use it for the returned result.
- https://github.com/woocommerce/woocommerce-android/pull/2760/files#r474648485
- Added input validation for the url/name in the file details.

**Screenshots**
<img width="342" alt="Screen Shot 2020-08-21 at 23 14 20" src="https://user-images.githubusercontent.com/1657201/90977586-6eed4d00-e53e-11ea-9a78-678cd6575a7e.png">  ![Screen Shot 2020-08-24 at 11 31 38](https://user-images.githubusercontent.com/1657201/91034795-7cb5d780-e5fd-11ea-9cc3-cb70611a8865.png)
<img width="342" alt="Screen Shot 2020-08-21 at 23 15 05" src="https://user-images.githubusercontent.com/1657201/90977584-6dbc2000-e53e-11ea-949a-51c7eda2a740.png"> <img width="342" alt="Screen Shot 2020-08-21 at 23 15 14" src="https://user-images.githubusercontent.com/1657201/90977583-6dbc2000-e53e-11ea-879e-ea0da9dd56ff.png">
![Screen Shot 2020-08-24 at 11 31 47](https://user-images.githubusercontent.com/1657201/91034855-93f4c500-e5fd-11ea-8780-417cecf37294.png) <img width="342" alt="Screen Shot 2020-08-21 at 23 17 07" src="https://user-images.githubusercontent.com/1657201/90977581-6b59c600-e53e-11ea-876d-2f586c0f411e.png">


**Testing**

**Test 1: Product not downloadable (or downloadable files is empty)**
1. Open product details.
2. Click on "Add more details"
3. Select "Downloadable files"
4. Make sure that a bottom sheet is displayed listing different sources to pick a file.
5. Test different scenarios, and confirm that when picking a file from device or camera, a progress dialog is shown while uploading the file.
6. Edit details of the file in the details screen, and save.
7. Confirm that the file was added.

**Test 2: Product has downloadable files**
1. Open product details.
2. Click on "downloadable files".
3. Click on "Add file"
4. same as steps 4-7 from test 1.
 
**Test 3: Input validation**
Make sure that input validation respects the following cases:
- URL empty => show the error
- URL non empty but doesn’t contain a path (something like: http://testurl.com/):
           - Name is empty => show an error that the name is missing.
           - Name non empty => input is valid, enable the done button.
- URL non empty and have a path (http://testurl.com/file.jpg) => input is valid regardless of the content of the name field.
- URL has to be a valid URL, if the user doesn’t enter a protocol, we’ll add “http” implicitly.

hichamboushaba committed:
Add fragment to handle files addition

hichamboushaba committed:
Allow adding from Wordpress media library

hichamboushaba committed:
fix tests

hichamboushaba committed:
Add support to pick files from file picker and camera

hichamboushaba committed:
Add support for adding files from the downloads fragment

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/wpmediapicker/WPMediaPickerFragment.kt:
```diff
@@ -137,10 +140,10 @@ class WPMediaPickerFragment : BaseFragment(), WPMediaGalleryListener, BackPressL
+                navArgs.requestCode,
+                bundle,
+                R.id.nav_host_fragment_main,
+                findNavController().previousBackStackEntry?.destination?.id ?: R.id.productDetailFragment
```
Having the R.id.productDetailFragment static here meant that we always popstack all the way to the ProductDetailFragment, which causes an issue after adding a file from the ProductDownloadsFragment.
I updated it to popstack to just the previous fragment, and updated ProductImagesFragment to intercept the result of picking images.
And this improved also the UX as the flow is similar to picking from camera/device now, check below gifs for the new behavior:

Before:
![wpmedia_pick_file_before](https://user-images.githubusercontent.com/1657201/90976698-f5059580-e536-11ea-8140-9890ee7a6513.gif)

After:
![wpmedia_pick_file_after](https://user-images.githubusercontent.com/1657201/90976702-fcc53a00-e536-11ea-9427-dd40ccca4c8d.gif)

anitaa1990 replied:
Nice <3

hichamboushaba committed:
Add validation test cases

hichamboushaba committed:
Code cleanup

hichamboushaba committed:
Show the downloadable files menu only if M4 feature is enabled

hichamboushaba committed:
Create FilePickerChooserView to be reusable for picking files scenarios

anitaa1990 reviewed (changes requested):
Nice work on this @hichamboushaba!! I just have a few comments after testing:

- I have a product with the downloadable flag switched on and the product does not have any downloadable files so it should be displayed in the bottom sheet right? I am not seeing that though.
- When choosing files from a device, we have the option to select multiple files but we only want to support single selection right?
- When app is resumed from background and `do not keep activities` is enabled, the added file is still displayed but the “Update” menu button is not, so there is no way to update. I’m guessing this has something to do with the fix you made on `develop`? If so, we can leave it as is :) 
- Can we update the progress dialog message when uploading a file? It says “Loading”, which feels incorrect. Maybe uploading file? 
- I see from the designs that we also have the option to manually enter file url and name in thee bottom sheet. Is that being done in another PR?

<img width="447" alt="Screenshot 2020-08-24 at 10 42 53 AM" src="https://user-images.githubusercontent.com/22608780/91006345-bd5e2280-e5f6-11ea-9ac7-2cc3ec11d3bf.png">

- When clearing the contents of the url + name of an exiting file, the update menu button is disabled but there is no error message displayed.
<img src="https://user-images.githubusercontent.com/22608780/91006611-7886bb80-e5f7-11ea-9ae2-5981ac760103.gif" width="300"/>

- When the name is empty and the url does not have a path, I am not seeing the error message and the UPDATE menu button is displayed, when it should be disabled?
<img src="https://user-images.githubusercontent.com/22608780/91006813-0bbff100-e5f8-11ea-883a-35ee7c805c08.png" width="300"/>

hichamboushaba committed:
disable multi selection when picking file from device

hichamboushaba committed:
Copy downloads attributes when merging products

hichamboushaba committed:
Revert "Create FilePickerChooserView to be reusable for picking files scenarios"

This reverts commit 8ee07126

hichamboushaba committed:
Update file picker bottomsheet to match design, and allow adding files manually

hichamboushaba committed:
Avoid clearing edittext errors automatically after editing text

hichamboushaba committed:
Update url validation rules

hichamboushaba commented:
> * I have a product with the downloadable flag switched on and the product does not have any downloadable files so it should be displayed in the bottom sheet right? I am not seeing that though.

On my side I see the item in the bottom sheet, can you please send me the returned payload from the backend to try to understand why it's happening for you.

> * When choosing files from a device, we have the option to select multiple files but we only want to support single selection right?

You are right, it hould be fixed now.

> * When app is resumed from background and `do not keep activities` is enabled, the added file is still displayed but the “Update” menu button is not, so there is no way to update. I’m guessing this has something to do with the fix you made on `develop`? If so, we can leave it as is :)

On my side I saw a different behavior, the file was removed, and after checking I found that I missed updating the `mergeProduct` function, it's done on 7ba7c277655f2d9f806beb87504320223225726c.

> * Can we update the progress dialog message when uploading a file? It says “Loading”, which feels incorrect. Maybe uploading file?

@Garance91540 what do you think?

> * I see from the designs that we also have the option to manually enter file url and name in thee bottom sheet. Is that being done in another PR?

I totally missed this part, I thought it had the same design as the add photos dialog 🤦 .
I updated it, and removed the class FilePickerChooserView as it doesn't make sense now.

> * When clearing the contents of the url + name of an exiting file, the update menu button is disabled but there is no error message displayed.

The issue is that the `WCMaterialOutlinedEditTextView` class was clearing the error automatically every time the text is changed, and as we re-apply the error only if it changed, we weren't reapplying it.
I made some updates on the commit: c3c7f4026d5c9f447fe93af0e987419f8b3e4173 to avoid clearing the error automatically, my understanding is that this was needed for `onSkuChanged`, so i updated its logic to clear the error from the viewmodel. 

> * When the name is empty and the url does not have a path, I am not seeing the error message and the UPDATE menu button is displayed, when it should be disabled?

You are right, I updated the validation rules: c7d34a77f7f9e147496d590511c755893041a7c3

hichamboushaba committed:
Code cleanup

Garance91540 commented:
@hichamboushaba , thank you for working on this! Is it possible to send me some screenshots for Dark mode + tablet too?
A light mode, phone screencast would also be helpful to get a better understanding of the flow

> Can we update the progress dialog message when uploading a file? It says “Loading”, which feels incorrect. Maybe uploading file?
@Garance91540 what do you think?

"Uploading files" sounds good 👍

hichamboushaba commented:
> thank you for working on this! Is it possible to send me some screenshots for Dark mode + tablet too?
A light mode, phone screencast would also be helpful to get a better understanding of the flow

@Garance91540 please find below some screenshots, if you feel something is missing, let me know and I'll add it.

**Dark mode**
![Screen Shot 2020-08-24 at 16 26 41](https://user-images.githubusercontent.com/1657201/91066244-658cdf00-e629-11ea-8406-26ec3d40a9ca.png) ![Screen Shot 2020-08-24 at 16 26 46](https://user-images.githubusercontent.com/1657201/91066238-64f44880-e629-11ea-9a5b-692e6e489bff.png)
![Screen Shot 2020-08-24 at 16 33 05](https://user-images.githubusercontent.com/1657201/91066298-763d5500-e629-11ea-903d-63c07ebdb9bd.png) ![Screen Shot 2020-08-24 at 16 33 41](https://user-images.githubusercontent.com/1657201/91066234-632a8500-e629-11ea-96b8-4dc33e95ba9a.png)

**Tablet**
![Screen Shot 2020-08-24 at 16 40 30](https://user-images.githubusercontent.com/1657201/91066356-87866180-e629-11ea-91d7-55ad1f65a513.png)
![Screen Shot 2020-08-24 at 16 41 04](https://user-images.githubusercontent.com/1657201/91066360-89e8bb80-e629-11ea-9c2d-834909fd42c7.png)
![Screen Shot 2020-08-24 at 16 41 17](https://user-images.githubusercontent.com/1657201/91066364-8b19e880-e629-11ea-99b4-1e3d66e38e4d.png)

**Screencast**
![add_files](https://user-images.githubusercontent.com/1657201/91066417-9c62f500-e629-11ea-80b8-f7b4019a3262.gif)

hichamboushaba committed:
update loading translation, and do some cleanup

Garance91540 commented:
@hichamboushaba this is great, thank you for working on this!

anitaa1990 reviewed (changes requested):
> I have a product with the downloadable flag switched on and the product does not have any downloadable files so it should be displayed in the bottom sheet right? I am not seeing that though.
On my side I see the item in the bottom sheet, can you please send me the returned payload from the backend to try to understand why it's happening for you.

Ahh, I found out the confusion here. It looks like the `downloadable` flag should only be displayed for Simple products. Th product I mentioned yesterday was an external product but the `downloadable` flag was displayed so I switched it on. But I didn’t see the `Downloadable` files section in the bottom sheet. Can we make sure to display the flag only if thee product is `Simple`?

- When clicking on manually enter url, I noticed that the error message is displayed even when the file url has a path and a file name. See below screenshot:

<img src="https://user-images.githubusercontent.com/22608780/91163388-1f984f80-e6eb-11ea-94a8-e10ebc909651.gif" width="300"/>

- When validating a file, I understand that when there is no path to a file url and the file name is empty, we should display an error. But I noticed that when the file url has no path and I remove the name of the file, the error displayed is something like “Enter a valid file url”, which is really confusing. @garance can confirm here but it would make more sense to display the error in this case, on the file name, instead of the file url, since I made changes to the file name only and not the url.

<img src="https://user-images.githubusercontent.com/22608780/91163463-3a6ac400-e6eb-11ea-9647-95821c3586f9.gif" width="300"/>

hichamboushaba commented:
> Ahh, I found out the confusion here. It looks like the `downloadable` flag should only be displayed for Simple products. Th product I mentioned yesterday was an external product but the `downloadable` flag was displayed so I switched it on. But I didn’t see the `Downloadable` files section in the bottom sheet. Can we make sure to display the flag only if thee product is `Simple`?

Sure, I'll do it.

> * When clicking on manually enter url, I noticed that the error message is displayed even when the file url has a path and a file name. See below screenshot:

You have to specify a scheme for your url (http:// or https://), because otherwise the backend considers that the url is a relative path in the server, and returns an error that it doesn't exist:

```json
{
    "code": "product_invalid_download",
    "message": "The downloadable file <code>test.com/file.jpg</code> cannot be used as it does not exist on the server.",
    "data": {
        "status": 400
    }
}
```

> * When validating a file, I understand that when there is no path to a file url and the file name is empty, we should display an error. But I noticed that when the file url has no path and I remove the name of the file, the error displayed is something like “Enter a valid file url”, which is really confusing. @garance can confirm here but it would make more sense to display the error in this case, on the file name, instead of the file url, since I made changes to the file name only and not the url.

Yes, I asked about this in my p2 comment, but when I got the answer, it confirmed the initial validation 🙂 , so I assumed it's what we want.

hichamboushaba committed:
Hide the isDownloadable checkbox if product is not simple

anitaa1990 reviewed (commented):
> Ahh, I found out the confusion here. It looks like the downloadable flag should only be displayed for Simple products. Th product I mentioned yesterday was an external product but the downloadable flag was displayed so I switched it on. But I didn’t see the Downloadable files section in the bottom sheet. Can we make sure to display the flag only if thee product is Simple?

Works great now 🎉 

> When clicking on manually enter url, I noticed that the error message is displayed even when the file url has a path and a file name. See below screenshot:
You have to specify a scheme for your url (http:// or https://), because otherwise the backend considers that the url is a relative path in the server, and returns an error that it doesn't exist:

Ah, fair enough. But can we make sure to add the scheme implicitly? I seem to remember we do a similar logic during login. It's not likely that users will type the scheme when entering a url so it makes sense to add it ourselves. WDYT?

> When validating a file, I understand that when there is no path to a file url and the file name is empty, we should display an error. But I noticed that when the file url has no path and I remove the name of the file, the error displayed is something like “Enter a valid file url”, which is really confusing. @garance can confirm here but it would make more sense to display the error in this case, on the file name, instead of the file url, since I made changes to the file name only and not the url.
Yes, I asked about this in my p2 comment, but when I got the answer, it confirmed the initial validation 🙂 , so I assumed it's what we want.

Based on the discussion from yesterday, we are going to be adding the error to the filename. Is that correct?

hichamboushaba committed:
update input validation rules, and add http protocol implicitly if it's missing

hichamboushaba commented:
> Ah, fair enough. But can we make sure to add the scheme implicitly? I seem to remember we do a similar logic during login. It's not likely that users will type the scheme when entering a url so it makes sense to add it ourselves. WDYT?

> Based on the discussion from yesterday, we are going to be adding the error to the filename. Is that correct?

Both are done, please check the commit 5fbb670f346bb4ab6e09047b8b5f4273f63f0a3d

hichamboushaba committed:
Rename method

anitaa1990 reviewed (approved):
LGTM! Nice work @hichamboushaba 🥇
